waspc/packages/wasp-config/src/mappers.ts

```diff
@@ -0,0 +1,443 @@
+    auth,
+    client,
+    db,
+    emailSender,
+    jobs,
+    pages,
+    queries,
+    routes,
+    server,
+    websocket,
+  } = spec
+
+  const pageNames = Array.from(pages.keys())
+  const routeNames = Array.from(routes.keys())
+  const parseEntityRef = makeRefParser('Entity', entityNames)
+  const parsePageRef = makeRefParser('Page', pageNames)
+  const parseRouteRef = makeRefParser('Route', routeNames)
+
+  // TODO: Try to build the entire object at once
+  const decls: AppSpec.Decl[] = []
+
+  // TODO: Find a way to make sure you've covered everything in compile time
+  for (const [pageName, pageConfig] of pages.entries()) {
+    decls.push({
+      declType: 'Page',
+      declName: pageName,
+      declValue: mapPage(pageConfig),
+    })
+  }
+
+  for (const [routeName, routeConfig] of routes.entries()) {
+    decls.push({
+      declType: 'Route',
+      declName: routeName,
+      declValue: mapRoute(routeConfig, parsePageRef),
+    })
+  }
+
+  decls.push({
+    declType: 'App',
+    declName: app.name,
+    declValue: mapApp(
+      app.config,
+      parseEntityRef,
+      parseRouteRef,
+      auth,
+      server,
+      client,
+      db,
+      emailSender,
+      websocket
+    ),
+  })
+
+  for (const [actionName, actionConfig] of actions.entries()) {
+    decls.push({
+      declType: 'Action',
+      declName: actionName,
+      declValue: mapOperationConfig(actionConfig, parseEntityRef),
+    })
+  }
+
+  for (const [queryName, queryConfig] of queries.entries()) {
+    decls.push({
+      declType: 'Query',
+      declName: queryName,
+      declValue: mapOperationConfig(queryConfig, parseEntityRef),
+    })
+  }
+
+  for (const [apiName, apiConfig] of apis.entries()) {
+    decls.push({
+      declType: 'Api',
+      declName: apiName,
+      declValue: mapApiConfig(apiConfig, parseEntityRef),
+    })
+  }
+
+  for (const [jobName, jobConfig] of jobs.entries()) {
+    decls.push({
+      declType: 'Job',
+      declName: jobName,
+      declValue: mapJob(jobConfig, parseEntityRef),
+    })
+  }
+
+  for (const [
+    apiNamespaceName,
+    apiNamespaceConfig,
+  ] of apiNamespaces.entries()) {
+    decls.push({
+      declType: 'ApiNamespace',
+      declName: apiNamespaceName,
+      declValue: mapApiNamespace(apiNamespaceConfig),
+    })
+  }
+
+  for (const [crudName, crudConfig] of spec.cruds.entries()) {
+    decls.push({
+      declType: 'Crud',
+      declName: crudName,
+      declValue: mapCrud(crudConfig, parseEntityRef),
+    })
+  }
+
+  return decls
+}
+
+function mapOperationConfig(
+  config: User.QueryConfig,
+  parseEntityRef: RefParser<'Entity'>
+): AppSpec.Query
+function mapOperationConfig(
+  config: User.ActionConfig,
+  parseEntityRef: RefParser<'Entity'>
+): AppSpec.Action
+function mapOperationConfig(
+  config: User.ActionConfig | User.QueryConfig,
+  parseEntityRef: RefParser<'Entity'>
+): AppSpec.Action | AppSpec.Query {
+  // TODO: How to make sure I've destructured everything?
```

Comment:
Interesting, I was asking myslf the same asme thing while reading this. I guess we use `...rest` and then somehow say that its type has to be empty array or something like taht?

Reply:
I solved the root issue in another way, we discussed it in person: https://github.com/wasp-lang/wasp/pull/2393/files#diff-0e8a94a09bb3cfb7a0d8ee9d0e1779d2ae6e871002a542711326f56e68707d0eR243